flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/InputProcessorUtil.java

```diff
@@ -109,10 +109,17 @@ private static CheckpointBarrierHandler createCheckpointBarrierHandler(
 		switch (config.getCheckpointMode()) {
 			case EXACTLY_ONCE:
 				if (config.isUnalignedCheckpointsEnabled()) {
-					return new CheckpointBarrierUnaligner(
-						numberOfInputChannelsPerGate.toArray(),
-						channelStateWriter,
-						taskName,
+					return new AlternatingCheckpointBarrierHandler(
+						new CheckpointBarrierAligner(
+							taskName,
+							channelIndexToInputGate,
+							inputGateToChannelIndexOffset,
+							toNotifyOnCheckpoint),
+						new CheckpointBarrierUnaligner(
+							numberOfInputChannelsPerGate.toArray(),
+							channelStateWriter,
+							taskName,
+							toNotifyOnCheckpoint),
```

Comment:
There are two methods in `CheckpointedInputGate` which are implemented in not very object oriented design (I guess they should be re-implemented) that would stop working with this change (I hope some tests will fail because of that):
```
	public void spillInflightBuffers(
			long checkpointId,
			int channelIndex,
			ChannelStateWriter channelStateWriter) throws IOException {
		if (((CheckpointBarrierUnaligner) barrierHandler).hasInflightData(checkpointId, channelIndex)) {
			inputGate.getChannel(channelIndex).spillInflightBuffers(checkpointId, channelStateWriter);
		}
	}

	public CompletableFuture<Void> getAllBarriersReceivedFuture(long checkpointId) {
		return ((CheckpointBarrierUnaligner) barrierHandler).getAllBarriersReceivedFuture(checkpointId);
	}
```
I guess `hasInflightData()` and `getAllBarriersReceivedFuture()` should be pulled to `CheckpointBarrierHandler` interface (separate commit?)?

Reply:
Good point! 
Moving up the methods.